### DISCUSSION TITLE
Modbus client is not connected to the inverter

### DISCUSSION THREAD

**Discussion by johanvh83** [wlcrs/huawei_solar#334] [Setting up the integration]
### Describe the issue

I checked the discussion but nothing could help me and it seems nobody is very responsive on the discussions.
The integrations worked for months, almost 2 years, without problem. 
Since 2 days, I don't have any data anymore. Tried the almost I could but can't succeed.
When I ping the inverter, I get a reply and on FusionSolar all data is available.
I was a bit desperate and finally removed and re-installed the integration but after the popup with the IP address/port/ID it gives: 'Failed to connect' so I'm not able to check logs.
The inverter gets Wifi access from my router. In the past I connected by the local IP 192.168.100.117 to the inverter.
This option is not available from the drop-down 'how do you connect'. Is the setup changed?

The HA logs:

Logger: huawei_solar.huawei_solar
Source: /usr/local/lib/python3.10/site-packages/huawei_solar/huawei_solar.py:293
First occurred: 7:57:50 PM (12 occurrences)
Last logged: 8:11:00 PM

Modbus client is not connected to the inverter.
NoneType: None

and

This error originated from a custom integration.

Logger: backoff
Source: custom_components/huawei_solar/config_flow.py:114
Integration: Huawei Solar
First occurred: 7:58:21 PM (2 occurrences)
Last logged: 8:11:00 PM

Giving up _do_read(...) after 6 tries (huawei_solar.exceptions.ConnectionInterruptedException: Modbus client is not connected to the inverter.)

### Bescribe your Huawei Solar Setup

Inverter Type: SUN2000-4.6KTL-L1
Inverter Firmware version: V200R001C00SPC130
SDongle present: no
Power meter present: single phase
Battery: LUNA2000 10kWh
Battery Firmware version: V100R002C00SPC106

### How do you connect to the inverter?

Via the `SUN2000-<serial_no>` WiFi

### Upload your Diagnostics File

Drag & Drop your Diagnostics File here.

### Upload your relevant debug logs

```text
N/A
```

### Please confirm the following:

- [ ] The problem is still present in the latest release of this integration.
- [ ] I did not find an existing issue describing this problem.
- [ ] I did upload the diagnostics-file that I could retrieve from  the 'Devices & Services Page'
- [ ] I did upload the relevant debug logs (via 'Enable Debug Logging'-feature or by manually configuring HA logging)

**Comment by tremebundo** [comment]
I had the same problem a few days ago and it worked again reinstaling this pakage.
Anyway, I connect to the installer access point which I gess is not your case.

**Comment by johanvh83** [comment]
Hi @tremebundo, which package you installed? The package is intend for Home Assistant or for the inverter?
Based on this discussion, https://github.com/wlcrs/huawei_solar/issues/332, Huawei managed the way the inverter works.
I would believe I'm the only victim who used the WLAN communication for Home Assistant.
Is the communication port an alternative? https://github.com/wlcrs/huawei_solar/wiki/Connecting-to-the-inverter 
Terminals 1 and 2 should give serial communication, in not changed by Huawei ;-)

**Comment by tarmor1** [comment]
This is a common issue for me as well after HA updating or HA restarting.
What helps are additional restarts or integration reloading: Integrations > Huawei integration > (click on 3 dots icon) > reload.
But it takes many HA restarts or reloads once it hits the "sweet spot"... :)

**Comment by tremebundo** [comment]
Hi @johanvh83 , I mean this custom component.
But what I found strange is that the sensor names changed. So, I needed to reconfigure my energy panel.

**Comment by MariuszWoszczynski** [comment]
I also have a problem with connecting to inverter last days

**Comment by diegomontanaduarte** [comment]
Same issue here!

**Comment by johanvh83** [comment]
@tarmor1 & @tremebundo, how guys do you connect to the inverter, by Wifi? Do you have the Dongle?
According my information Huawei changed 'something' remotely and disabled Modbus by Wifi.
Seems there are 2 options now: Buy the Dongle to enable this feature again or use the serial output on the communication port. The last option required additional hardware. See https://github.com/wlcrs/huawei_solar/wiki/Connecting-to-the-inverter

**Reply by tarmor1** [reply]
> @tarmor1 & @tremebundo, how guys do you connect to the inverter, by Wifi? Do you have the Dongle? According my information Huawei changed 'something' remotely and disabled Modbus by Wifi. Seems there are 2 options now: Buy the Dongle to enable this feature again or use the serial output on the communication port. The last option required additional hardware. See https://github.com/wlcrs/huawei_solar/wiki/Connecting-to-the-inverter

As for me, I have the dongle connecting to my home WiFi (same network my other devices and HA is in). So the dongle sends data to Huawei cloud and HA/huawei_solar can access over wifi on modbus.

![image](https://user-images.githubusercontent.com/106139715/226706737-14c03d41-8805-482b-b4e4-a2f29ab9a134.png)



Once I got the inverter, I asked my installer to enable the Modbus as per instructions of this integration. I don't have installer access/rights myself.

Not sure if you have the full access for enabling or checking modbus status, maybe a valuable info would be to check with your installer for them to check what the Modbus status is for you now?

What I have definitely seen is that if Huawei cloud is down ("under maintenance"), then dongle flashed red (can't connect to cloud or whatever AND, HA/huawei_solar can't access my inverter either. Once this happened, I spent significant amount of time to debug and saw a clear pattern in my network side - inverter connected to Wifi for 30 seconds (get IP from DHCP, all that, connect, fail with outside network) and then disconnect for 60 sec and then repeat. Obviously this was issue for inverter data in HA as well. Not sure what Huawei cloud did answer to inverter, but nevertheless dongle tried to fix situation by repetitive connects/disconnects to my local network. This disconnect-reconnect lasted for hours. Once Huawei cloud started answering, all went good on HA side too and red on dongle turned green. At least that was my understanding of it. 

Maybe to check - does your official Huawei App show inverter status and all is OK? Maybe you have some cloud/account issue? Any chance to debug by the inverter lights? Eg, if dongle has red light, it means error.

Maybe can you check too - can you access via Telnet or some other low-level modbus thing?

**Reply by johanvh83** [reply]
Many thanks for your detailed answer. I don't have the Dongle, but even without the Dongle everything worked in the past.
Inverter was connected to internet as a client on my Wifi network. The same way HA read all data on the local IP address of my network 192.168.100.xxx. In the part there was no need to have the Dongle. Seems Huawei changed their strategy by disabling the Modbus data on the Wifi pas, as long as there is no Dongle installed now.

**Reply by tarmor1** [reply]
I'm wondering what has been the point of Wifi dongle at all if inverter itself was able to connect to Local Wifi...?
For the installer part I understand (creating AP on it's own).
So you might be right that's a strategy to sell dongles.
But I assume you are not alone in there and internet should be aware of it already..

I assume too you have tried the usual suspect: restart inverter :D

**Comment by diegomontanaduarte** [comment]
@johanvh83 even with new inverter firmware is still be possible to connect by using a WiFi USB stick on your Home Assistant server. In my case when I restarted Home Assistant works fine for some minutes and then disconnected from the inverter

**Comment by johanvh83** [comment]
So you believe the inverter still sends Modbus data over TCP (Wifi). Hopefully someone can clarify, maybe @wlcrs

**Comment by tremebundo** [comment]
Yes, I am connected right now to the wifi for the installer in the inverter. The one with fixed IP 192.168.1.200 and I get the data. I do not have any dongle or serial connection. And it has the latest firmware revision.
@diegomontanaduarte , when this problem appeared, I had to uninstall this component, install it again and configure to get it work again. I had the same symptoms as you. But since then, the sensor names have changed.

**Comment by johanvh83** [comment]
@wlcrs, so you are connnected to the Access Point from the inverter with SSID SUN2000-etc.
You have on the the HA side, kind of additional Wifi adapter to build a Wifi bridge?
How is your inverter connected to internet, or isn't it?
In my case, and I believe maybe also like most other users, HA and the inverter are in the same network, my main home network let's say.
The inverter is connected to the Wifi part of my home network.The installers Wifi isn't used. 
I used it this way because it was the easiest solution, without need of additional hardware.
It worked for about 1,5 year and since some days I'm not able to retrieve data this way. I removed your integration, re-installed but I get 'Failed to connect' during the setup process. I read that Huawei remotely disabled the Modbus by TCP feature. You should buy and install a dongle to be able to enable it again, or use the serial data output.

**Comment by AtomArne** [comment]
Had the same issue, updated to the newest dongle firmware and the newest HA version (Core and Supervisor), removed the Huawai Solar integration, reinstalled it, connected to the dongle thru FE on Port 502 Slave 1 and bingo.

My recommendation  is you contact your installer and make them change your user from "Owner" to "Installer" on your PV. From the webinterface (with the installer role) you can easily pull and upgrade your devices with the newest firmware thru the "Update Managment" menu:

![image](https://user-images.githubusercontent.com/20305426/227322321-7a9e6f3f-bd4d-4d71-9b51-0ba5e2b279ff.png)

**Comment by cwoerndl** [comment]
Hey, does anyone can support with this issue?

I'm currently setting up a new HA OS as my current one is more like a sandbox and running in a docker container.
My Huawei Inverter has a Dongle connected which is connected to my network with a fixed IP.
Version "old" HA: 2023.12.3 (don't want to update as this is working fine)
Version "new" HA: 2023.12.4
integration version on both HA: 1.2.9 (tried this as this is working on my "old" HA, but also tried 1.3.1 and had the same issue)

On my "old" HA running in docker the integration is working fine, I'm getting all data and had no such error within the last 7 months (since I configured the integration the first time).

Now on my "new" HA I started adding all integration I'd need, but when trying to configure this integration with IP, port (502 or 6607 tried both) and slave ID's "1" as the dongle is connected, I'm always getting this warnings and error:

----
This error originated from a custom integration.

Logger: huawei_solar.huawei_solar
Source: custom_components/huawei_solar/config_flow.py:115
Integration: Huawei Solar (documentation, issues)
First occurred: 11:48:06 (1 occurrences)
Last logged: 11:48:06

Could not read register value, has another device interrupted the connection?
---
Logger: huawei_solar.huawei_solar
Source: /usr/local/lib/python3.11/site-packages/huawei_solar/huawei_solar.py:348
First occurred: 09:01:19 (8 occurrences)
Last logged: 11:48:10

Modbus client is not connected to the inverter.
NoneType: None
---
Logger: huawei_solar.huawei_solar
Source: /usr/local/lib/python3.11/site-packages/huawei_solar/huawei_solar.py:184
First occurred: 11:48:54 (1 occurrences)
Last logged: 11:48:54

Aborting client creation due to error.
Traceback (most recent call last):
  File "/usr/local/lib/python3.11/site-packages/huawei_solar/huawei_solar.py", line 178, in create
    await huawei_solar._initialize()
  File "/usr/local/lib/python3.11/site-packages/huawei_solar/huawei_solar.py", line 107, in _initialize
    self.time_zone = (await self.get(rn.TIME_ZONE)).value
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/huawei_solar/huawei_solar.py", line 241, in get
    return (await self.get_multiple([name], slave))[0]
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/huawei_solar/huawei_solar.py", line 276, in get_multiple
    response = await self._read_registers(registers[0].register, total_length, slave)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/huawei_solar/huawei_solar.py", line 398, in _read_registers
    return await _do_read()
           ^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/backoff/_async.py", line 151, in retry
    ret = await target(*args, **kwargs)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/backoff/_async.py", line 151, in retry
    ret = await target(*args, **kwargs)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/huawei_solar/huawei_solar.py", line 351, in _do_read
    response = await self._client.read_holding_registers(
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/pymodbus/client/base.py", line 207, in async_execute
    raise ModbusIOException(
pymodbus.exceptions.ModbusIOException: Modbus Error: [Input/Output] ERROR: No response received after 3 retries


+++

at the same time the integration on my "old" HA is still working fine....

Can anyone support?


thx & br

**Reply by cwoerndl** [reply]
I found the source of this issue.
I had 2 HA instances running at the same time and it is only possible to run a working integration in 1 HA at the same time (I assume, as after shutting down the old HA setting up the integration in the new one worked out fine without any issues)

br

**Reply by Roving-Ronin** [reply]
This is as the sDongle / inverter only supports ONE modbus connection at any time. If you want to have multiple instances of WLCRS (or any other application) querying the inverter(s) then you must run a modbus proxy. There is a modbus proxy available in HA via HACS, but look at someother threads on here, as some people have run into issues that the proxy takes too long to start up, with WLCRS integration timing out in the meantime. IIRC one thread mentions how to hack the .py files to change the timeout value.

Personally just hanging out for the new EMMA appliance that replaces the sDongle / Logger / DTSU-666H meter and that iirc supports multiple modbus-tcp client connects (assuming WLCRS works with it, if it has the same modbus registers or if updates to WLCRS will need to be released).
